clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/auth/DockerBearerTokenService.groovy

```diff
@@ -0,0 +1,206 @@
+          value = header
+          header = ""
+        } else {
+          value = header.substring(0, valueEnd)
+          header = header.substring(valueEnd + 1)
+        }
+      }
+
+      if (key.equalsIgnoreCase(realmKey)) {
+        def url = new URL(value)
+        result.realm = url.protocol + "://" + url.authority
+        result.path = url.path
+        if (result.path.length() > 0 && result.path[0] == "/") {
+          result.path = result.path.substring(1)
+        }
+      } else if (key.equalsIgnoreCase(serviceKey)) {
+        result.service = value
+      } else if (key.equalsIgnoreCase(scopeKey)) {
+        result.scope = value
+      }
+    }
+
+    if (!result.realm) {
+      throw new DockerRegistryAuthenticationException('Www-Authenticate header must provide "realm" parameter.')
+    }
+    if (!result.service) {
+      throw new DockerRegistryAuthenticationException('Www-Authenticate header must provide "service" parameter.')
+    }
+    if (!result.scope) {
+      throw new DockerRegistryAuthenticationException('Www-Authenticate header must provide "scope" parameter.')
+    }
+
+    return result
+  }
+
+  private getTokenService(String realm) {
+    def tokenService = realmToService.get(realm)
+
+    if (tokenService == null) {
+      def builder = new RestAdapter.Builder().setEndpoint(realm).setLogLevel(RestAdapter.LogLevel.BASIC).build()
+      tokenService = builder.create(TokenService.class)
+      realmToService[realm] = tokenService
+    }
+
+    return tokenService
+  }
+
+  public DockerBearerToken getToken(String repository) {
+    return cachedTokens[repository]
+  }
+
+  public DockerBearerToken getToken(String repository, List<Header> headers) {
+    String authenticate = null
+
+    headers.forEach { header ->
+      if (header.name == "Www-Authenticate") {
+        authenticate = header.value
+      }
+    }
+
+    if (!authenticate) {
+      throw new DockerRegistryAuthenticationException("Www-Authenticate header must be supplied for Docker v2 registry token authentication.")
+    }
+
+    def authenticateDetails
+    try {
+      authenticateDetails = parseBearerAuthenticateHeader(authenticate)
+    } catch (Exception e) {
+      throw new DockerRegistryAuthenticationException("Failed to retrieve token: ${e.message}")
+    }
+
+    def tokenService = getTokenService(authenticateDetails.realm)
+    def token
+    if (basicAuthenticate) {
+      token = tokenService.getToken(authenticateDetails.path, authenticateDetails.service, authenticateDetails.scope, basicAuthenticate)
+    }
+    else {
+      token = tokenService.getToken(authenticateDetails.path, authenticateDetails.service, authenticateDetails.scope)
+    }
+
+    cachedTokens[repository] = token
+    return token
+  }
+
+  private interface TokenService {
+    @GET("/{path}")
+    @Headers([
+      "User-Agent: Spinnaker-Clouddriver",
+      "Docker-Distribution-API-Version: registry/2.0"
+    ]) // TODO(lwander) get clouddriver version #
```

Comment:
Can probably do this in a followon PR (and promote the clouddriver version stuff out of the google package).